components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4780,11 +4927,66 @@ where
         self.register_split_region_check_tick();
     }
 
+    // generate bucket range list to run split-check (to further split buckets)
+    fn gen_bucket_range_for_update(&self) -> Option<Vec<BucketRange>> {
+        if !self.ctx.coprocessor_host.cfg.enable_region_bucket {
+            return None;
+        }
+        let region_buckets = self.fsm.peer.region_buckets.as_ref()?;
+        let stats = &region_buckets.stats;
+        let keys = &region_buckets.meta.keys;
+
+        let empty_last_keys = vec![];
+        let empty_last_stats = metapb::BucketStats::default();
+        let (last_keys, last_stats, stats_reset) = self
+            .fsm
+            .peer
+            .last_region_buckets
+            .as_ref()
+            .map(|b| {
+                (
+                    &b.meta.keys,
+                    &b.stats,
+                    region_buckets.last_report_time != b.last_report_time,
+                )
+            })
+            .unwrap_or((&empty_last_keys, &empty_last_stats, false));
+
+        let mut bucket_ranges = vec![];
+        let mut j = 0;
+        for i in 0..stats.write_bytes.len() {
+            let mut diff_in_bytes = stats.write_bytes[i];
+            while j < last_keys.len() && keys[i] > last_keys[j] {
+                j += 1;
+            }
+            if j < last_keys.len() && keys[i] == last_keys[j] {
+                if !stats_reset {
+                    diff_in_bytes -= last_stats.write_bytes[j];
+                }
+                j += 1;
+            }
+
+            // if the bucket's write_bytes exceed half of the configured region_bucket_size,
+            // add it to the bucket_ranges for checking update
+            let bucket_update_diff_size_threshold =
+                self.ctx.coprocessor_host.cfg.region_bucket_size.0 / 2;
+            if diff_in_bytes >= bucket_update_diff_size_threshold {
+                if i == stats.write_bytes.len() - 1 {
+                    bucket_ranges.push(BucketRange(keys[i].clone(), vec![]));
+                } else {
+                    bucket_ranges.push(BucketRange(keys[i].clone(), keys[i + 1].clone()));
+                }
```

Comment:
`keys`'s length == `write_bytes`'s length + 1, and you shouldn't use `vec![]` as the end key because checker doesn't handle it properly.
```suggestion
                bucket_ranges.push(BucketRange(keys[i].clone(), keys[i + 1].clone()));
```

Reply:
> `keys`'s length == `write_bytes`'s length + 1, and you shouldn't use `vec![]` as the end key because checker doesn't handle it properly.

Good catch. Updated the checker to handle vec![] properly.